desktop/src/gui/locale.rs

```diff
@@ -0,0 +1,131 @@
+        });
+    reorder_bidi(locale, text)
+}
+
+pub fn optional_text(locale: &LanguageIdentifier, id: &str) -> Option<String> {
+    TEXTS
+        .lookup_single_language::<&str>(locale, id, None)
+        .inspect_err(|e| tracing::trace!("Error looking up text: {e}"))
+        .ok()
+        .map(|s| reorder_bidi(locale, s.into()).into_owned())
+}
+
+pub fn available_languages() -> Vec<&'static LanguageIdentifier> {
+    let mut result: Vec<_> = TEXTS.locales().collect();
+    result.sort();
+    result
+}
+
+#[allow(dead_code)]
+pub fn text_with_args(
+    locale: &LanguageIdentifier,
+    id: &'static str,
+    args: &HashMap<Cow<'static, str>, FluentValue>,
+) -> Cow<'static, str> {
+    let text = TEXTS
+        .try_lookup_with_args(locale, id, args)
+        .map(Cow::Owned)
+        .unwrap_or_else(|| {
+            tracing::error!("Unknown desktop text id '{id}'");
+            Cow::Borrowed(id)
+        });
+    reorder_bidi(locale, text)
+}
+
+/// Reorder BiDi text so that RTL text runs are reversed.
+/// The default direction is based on the locale.
+///
+/// TODO This is stupid, but it at least allows people to read RTL languages
+///   somewhat correctly. Remove it when egui starts supporting RTL scripts.
+fn reorder_bidi(locale: &LanguageIdentifier, text: Cow<'static, str>) -> Cow<'static, str> {
+    let level = if is_rtl(locale) {
+        unicode_bidi::Level::rtl()
+    } else {
+        unicode_bidi::Level::ltr()
+    };
+
+    let bidi_info = unicode_bidi::BidiInfo::new(&text, Some(level));
+    if !unicode_bidi::level::has_rtl(&bidi_info.levels) {
+        // Fast path: no RTL text = no reordering
+        return text;
+    }
+
+    // At this point we know something has to be reversed.
+    let mut reordered_text = String::with_capacity(text.len());
+    for para in &bidi_info.paragraphs {
+        let (levels, runs) = bidi_info.visual_runs(para, para.range.clone());
+        for run in runs {
+            if levels[run.start].is_rtl() {
+                reordered_text.extend(text[run].chars().rev().map(mirror_char));
+            } else {
+                reordered_text.push_str(&text[run]);
+            }
+        }
+    }
+
+    reordered_text.into()
+}
+
+/// Check is the given locale has a RTL script.
+///
+/// That's a very simplistic way of detecting RTL,
+/// but it's enough for our use.
+fn is_rtl(langid: &LanguageIdentifier) -> bool {
+    matches!(
+        langid.language.as_str(),
+        "ar" | "he" | "fa" | "ur" | "ps" | "syr" | "dv"
+    )
+}
+
+/// Mirror the given character.
+///
+/// It's used as a very simple alternative to glyph mirroring, which we do not
+/// support yet and allows basic characters to be rendered properly in RTL.
+fn mirror_char(c: char) -> char {
+    match c {
+        '(' => ')',
+        ')' => '(',
+        '[' => ']',
+        ']' => '[',
+        '{' => '}',
+        '}' => '{',
+        '<' => '>',
+        '>' => '<',
+        _ => c,
+    }
+}
```

Comment:
FWIW, there's also https://docs.rs/unicode-bidi-mirroring/0.4.0/unicode_bidi_mirroring/fn.get_mirrored.html.

Reply:
I'm not sure if it's worth adding another dependency, as the set of strings we mirror is finite (we do that only for translated strings), but I'll add it if you think it's better \:D

Reply:
I agree, either way is fine for me.